spec/ics-027-interchain-account/README.md

```diff
@@ -0,0 +1,285 @@
+  abortTransactionUnless(counterpartyPortIdentifier === "interchain-account")  
+}  
+```
+
+```typescript  
+function onChanOpenAck(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier,  
+  version: string) {  
+  // version not used at present  
+  abortTransactionUnless(version === "")  
+  // port has already been validated  
+}  
+```
+
+```typescript  
+function onChanOpenConfirm(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier) {  
+  // accept channel confirmations, port has already been validated  
+}  
+```
+
+```typescript  
+function onChanCloseInit(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier) {  
+  // no action necessary  
+}  
+```
+
+```typescript  
+function onChanCloseConfirm(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier) {  
+  // no action necessary  
+}  
+```
+
+### **Packet relay**  
+
+In plain English, between chains `A` and `B`. It will describe only the case that chain A wants to register an Interchain account on chain B and control it. Moreover, this system can also be applied the other way around.  
+
+```typescript  
+function onRecvPacket(packet: Packet): bytes {  
+  if (packet.data is RunTxPacketData) {  
+    const tx = deserializeTx(packet.data.txBytes)
+    abortTransactionUnless(authenticateTx(tx) == true)
+    return runTx(tx)
+  }  
+      
+  if (packet.data is RegisterIBCAccountPacketData) {  
+    RegisterIBCAccountPacketData data = packet.data  
+    identifier = "{packet/sourcePort}/{packet.sourceChannel}"  
+    const address = generateAddress(identifier, packet.salt)
+    createAccount(address)
+    // Return generated address.
+    return address
+  }   
+      
+  return 0x  
+}  
+```
+
+```typescript  
+function onAcknowledgePacket(  
+  packet: Packet,  
+  acknowledgement: bytes) {  
+  if (packet.data is RegisterIBCAccountPacketData) {
+    if (acknowledgement !== 0x) {
+      identifier = "{packet/sourcePort}/{packet.sourceChannel}"
+      onAccountCreated(identifier, acknowledgement)
+    }
+  }
+  if (packet.data is RunTxPacketData) {
+    identifier = "{packet/destPort}/{packet.destChannel}"  
+    onTxSucceeded(identifier: Identifier, packet.data.tyBytes)
+  }
+}  
+```
+
+```typescript  
+function onTimeoutPacket(packet: Packet) {  
+  // Receiving chain should handle this event as if the tx in packet has failed  
+  if (packet.data is RunTxPacketData) {
+    identifier = "{packet/destPort}/{packet.destChannel}"  
+    onTxFailed(identifier: Identifier, packet.data.tyBytes)
+  }
+}  
+```
+
+```typescript  
+function onTimeoutPacketClose(packet: Packet) {  
+  // nothing is necessary  
+}  
+```
+
+## **Backwards Compatibility**  
+
+Not applicable.
+
+## **Forwards Compatibility**  
+
+Not applicable.
+
+## **Example Implementation**  
+
+Pseudocode for cosmos-sdk: https://github.com/everett-protocol/everett-hackathon/tree/master/x/interchain-account
```

Comment:
It would also be helpful to reason about how this would work on Ethereum, which has a sufficiently programmable system that it should be able to implement this spec.

Reply:
I think that it can be implemented in a smart contract level with using CALL, DELEGATECALL, CALLCODE opcode that are used to delegate call to some function in callee. I will research more and try to implement pseudocode in Solidity.